test/lit/passes/type-merging.wast

```diff
@@ -537,6 +562,60 @@
   )
 )
 
+;; Check that we fully optimize a type graph that requires multiple iterations
+;; of supertype and sibling merging.
+(module
+  (rec
+    ;; These will get merged in the initial supertype merging stage.
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $B' (struct (field (ref $A))))
+
+    ;; CHECK:       (type $C (struct_subtype (field (ref $A)) (field i32) $B'))
+
+    ;; CHECK:       (type $D' (struct_subtype (field (ref $A)) (field i32) (field i32) $C))
+
+    ;; CHECK:       (type $A (struct ))
+    (type $A (struct))
+    (type $A' (struct_subtype $A))
+
+    ;; These siblings will be merged only after $a and $a' are merged.
+    (type $B (struct (ref $A)))
+    (type $B' (struct (ref $A')))
+
+    ;; These will get merged only after $b and $b' are merged.
+    (type $C (struct_subtype (ref $A) i32 $B))
+    (type $C' (struct_subtype (ref $A') i32 $B'))
+
+    ;; These will get merged only after $c and $c' are merged.
+    (type $D (struct_subtype (ref $A) i32 i32 $C))
+    (type $D' (struct_subtype (ref $A') i32 i32 $C'))
+  )
+
+  ;; CHECK:       (type $none_=>_none (func))
+
+  ;; CHECK:      (func $foo (type $none_=>_none)
+  ;; CHECK-NEXT:  (local $a (ref null $A))
+  ;; CHECK-NEXT:  (local $a' (ref null $A))
+  ;; CHECK-NEXT:  (local $b (ref null $B'))
+  ;; CHECK-NEXT:  (local $b' (ref null $B'))
+  ;; CHECK-NEXT:  (local $c (ref null $C))
+  ;; CHECK-NEXT:  (local $c' (ref null $C))
+  ;; CHECK-NEXT:  (local $d (ref null $D'))
+  ;; CHECK-NEXT:  (local $d' (ref null $D'))
+  ;; CHECK-NEXT:  (nop)
```

Comment:
I'm pretty sure it is deterministic, but the actual order depends on the order in which we collect types on the module, the order in which DFA minimization happens to emit elements in its output partitions, and the implementation details of our topological sort utility, so there is definitely no intention or rationale for a particular ordering.

In general I think it would be fine if future changes to this pass produced different orderings like this change does.

Reply:
sgtm. Any ordering is fine as long as it is deterministic. If not the fuzzer should find that.